WordPress/Classes/ViewRelated/Post/PostEditorState.swift

```diff
@@ -0,0 +1,292 @@
+    }
+
+    func updated(publishDate: Date?) {
+        let updatedState = editorState.updated(publishDate: publishDate, context: self)
+        guard type(of: editorState) != type(of: updatedState) else {
+            return
+        }
+
+        editorState = updatedState
+    }
+
+    func updated(hasContent: Bool) {
+        self.hasContent = hasContent
+    }
+
+    func updated(isBeingPublished: Bool) {
+        self.isBeingPublished = isBeingPublished
+    }
+
+    var action: PostEditorAction {
+        return editorState.action
+    }
+
+    var publishButtonText: String {
+        return editorState.action.publishActionLabel
+    }
+
+    var publishVerbText: String {
+        return editorState.action.publishingActionLabel
+    }
+
+    var isPostPostShown: Bool {
+        return editorState.action.isPostPostShown
+    }
+
+    // TODO: Replace with a method to bring label and such back for secondary publish button
+//    var isSecondaryPublishButtonShown: Bool {
+//        return editorState.isSecondaryPublishButtonShown(context: self)
+//    }
+
+    var isPublishButtonEnabled: Bool {
+        return publishActionAllowed
+    }
+
+}
+
+fileprivate class PostEditorStatePublish: PostEditorActionState {
+    var action: PostEditorAction {
+        return .publish
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        switch postStatus {
+        case .draft where context.originalPostStatus == .publish:
+            return PostEditorStateUpdate()
+        case .draft:
+            return PostEditorStateSave()
+        default:
+            return self
+        }
+    }
+
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState {
+        guard let publishDate = publishDate else {
+            return self
+        }
+
+        if isFutureDated(publishDate) {
+            return PostEditorStateSchedule()
+        }
+
+        return self
+    }
+}
+
+fileprivate class PostEditorStateSave: PostEditorActionState {
+    var action: PostEditorAction {
+        return .save
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        switch postStatus {
+        case .publish:
+            return PostEditorStateUpdate()
+        default:
+            return self
+        }
+    }
+
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState {
+        guard let publishDate = publishDate else {
+            return self
+        }
+
+        if isFutureDated(publishDate) {
+            return PostEditorStateSchedule()
+        }
+
+        return self
+    }
+}
+
+fileprivate class PostEditorStateSchedule: PostEditorActionState {
+    var action: PostEditorAction {
+        return .schedule
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
```

Comment:
Presumably this is why I can't seem to get the button out of Schedule state once I've entered it?